File: Backend/Backend/Models/ListaPrecio.cs
namespace Backend.Models;

/// <summary>
/// Politica de precios: Mayorista, Minorista, Bodega. Un cliente compra con la
/// lista que tenga asignada; si no tiene, con la predeterminada.
/// </summary>
public class ListaPrecio
{
    public int Id { get; set; }
    public string Nombre { get; set; } = string.Empty;
    public string? Descripcion { get; set; }

    /// <summary>La que se usa cuando el cliente no tiene lista propia.</summary>
    public bool EsPredeterminada { get; set; }

    public bool Activo { get; set; } = true;
    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;

    public List<PrecioProducto> Precios { get; set; } = [];
}

/// <summary>
/// Precio de una presentacion dentro de una lista.
///
/// El precio se pone POR PRESENTACION, no por producto: asi el saco sale mas
/// barato por kilo que vender el kilo suelto, que es como funciona el negocio.
///
///   Azucar, lista Minorista
///     Kilo         S/ 4.50   -> 4.50 el kilo
///     Bolsa 5 kg   S/ 21.00  -> 4.20 el kilo
///     Saco 50 kg   S/ 195.00 -> 3.90 el kilo
///
/// <see cref="CantidadMinima"/> permite ademas un escalon dentro de la misma
/// lista: desde 10 sacos, cada saco a S/ 190.
/// </summary>
public class PrecioProducto
{
    public int Id { get; set; }

    public int ListaPrecioId { get; set; }
    public ListaPrecio? ListaPrecio { get; set; }

    public int PresentacionId { get; set; }
    public ProductoPresentacion? Presentacion { get; set; }

    /// <summary>Precio de UNA presentacion completa: un saco, una caja, un kilo.</summary>
    public decimal Precio { get; set; }

    /// <summary>
    /// Desde cuantas presentaciones aplica este precio. 1 es el precio normal;
    /// un registro con 10 es el escalon por volumen.
    /// </summary>
    public decimal CantidadMinima { get; set; } = 1m;

    public bool Activo { get; set; } = true;
    public DateTime FechaActualizacion { get; set; } = DateTime.UtcNow;
}

File: titanic/lib/features/facturacion/datos/lista_precio.dart
/// Una lista de precios: catalogo de a cuanto se vende cada presentacion.
class ListaPrecio {
  const ListaPrecio({
    required this.id,
    required this.nombre,
    this.descripcion,
    required this.esPredeterminada,
    required this.activo,
    required this.precios,
  });

  final int id;
  final String nombre;
  final String? descripcion;
  final bool esPredeterminada;
  final bool activo;

  /// Cuantos precios tiene cargados.
  final int precios;

  String get buscable => '$nombre ${descripcion ?? ''}'.toLowerCase();

  factory ListaPrecio.desdeJson(Map<String, dynamic> json) => ListaPrecio(
    id: json['id'] as int,
    nombre: json['nombre'] as String? ?? '',
    descripcion: json['descripcion'] as String?,
    esPredeterminada: json['esPredeterminada'] as bool? ?? false,
    activo: json['activo'] as bool? ?? true,
    precios: json['precios'] as int? ?? 0,
  );
}

/// El precio de una presentacion dentro de una lista, con su escalon por
/// volumen (desde cuantas unidades aplica).
class Precio {
  const Precio({
    required this.id,
    required this.listaPrecioId,
    required this.listaPrecio,
    required this.presentacionId,
    required this.presentacion,
    required this.productoId,
    required this.producto,
    required this.precio,
    required this.cantidadMinima,
    required this.precioUnidadBase,
    required this.unidadBase,
    required this.activo,
    required this.fechaActualizacion,
  });

  final int id;
  final int listaPrecioId;
  final String listaPrecio;

  final int presentacionId;
  final String presentacion;

  final int productoId;
  final String producto;

  final double precio;

  /// Desde cuantas presentaciones aplica. 1 es el precio normal.
  final double cantidadMinima;

  /// Precio / factor: deja comparar el saco contra el kilo suelto.
  final double precioUnidadBase;
  final String unidadBase;

  final bool activo;
  final DateTime fechaActualizacion;

  String get buscable => '$producto $presentacion'.toLowerCase();

  factory Precio.desdeJson(Map<String, dynamic> json) => Precio(
    id: json['id'] as int,
    listaPrecioId: json['listaPrecioId'] as int,
    listaPrecio: json['listaPrecio'] as String? ?? '',
    presentacionId: json['presentacionId'] as int,
    presentacion: json['presentacion'] as String? ?? '',
    productoId: json['productoId'] as int,
    producto: json['producto'] as String? ?? '',
    precio: (json['precio'] as num?)?.toDouble() ?? 0,
    cantidadMinima: (json['cantidadMinima'] as num?)?.toDouble() ?? 1,
    precioUnidadBase: (json['precioUnidadBase'] as num?)?.toDouble() ?? 0,
    unidadBase: json['unidadBase'] as String? ?? '',
    activo: json['activo'] as bool? ?? true,
    fechaActualizacion: DateTime.parse(json['fechaActualizacion'] as String),
  );
}
